core/src/main/java/io/grpc/Metadata.java

```diff
@@ -668,19 +668,19 @@ public final String name() {
      * Returns true if the two objects are both Keys, and their names match (case insensitive).
      */
     @Override
-    public boolean equals(Object o) {
+    public final boolean equals(Object o) {
       if (this == o) {
         return true;
       }
-      if (o == null || getClass() != o.getClass()) {
+      if (!(o instanceof Key)) {
```

Comment:
I think the original was correct.  We only want AsciiKey to be equal to AsciiKey, and BinaryKey to be a BinaryKey

Reply:
The new version would have functioned correctly, but was more brittle, since it relied on binary keys having -bin in the name and ascii keys not.